### ISSUE TITLE
std.ChildProcess can deadlock if both stdout and stderr are long and interleaved

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#6343]
Related: #938 but this needs to work with blocking I/O too.

https://github.com/ziglang/zig/blob/d073836894df8b9fb56f24f577a51dd09a85a932/lib/std/child_process.zig#L216-L220

This is a flaw in the API design because this example code above will be duplicated many times in users of this API. We need to eliminate this footgun.

**Comment by LemonBoy** [comment]
A simple test case to show the deadlock on Linux:

```zig
const std = @import("std");

pub fn main() !void {
     const allocator = std.heap.page_allocator;

     const args = try std.process.argsAlloc(allocator);

     if (args.len != 1) {
         // Enough to saturate the pipe buffer
         try std.io.getStdErr().writeAll("A" ** 65536);
         try std.io.getStdErr().writeAll("A" ** 65536);
         return;
     }

     const child = try std.ChildProcess.exec(.{
         .allocator = allocator,
         // Re-exec this program and fill the stderr pipe
         .argv = &[_][]const u8{ "foo", "-nonexistentflag" },
         .max_output_bytes = 1_000_000,
     });

     std.debug.print("O {}", .{child.stdout});
     std.debug.print("E {}", .{child.stderr});
}
```

From my POV the API is great as-is, if the user asks for stdout/stderr to be a pipe they must be ready to handle both streams at the same time using async IO/threads or whatever they prefer. The problem here is that some consumers, such as the ones in the Zig codebase marked with `todo`, only need a way to execute some command and read all the output.
To accommodate such uses of `ChildProcess` I'd propose to:
* Add the polling logic in `exec`. This is the blessed function that abstracts the complexity away from the user, reading stdin/stdout by hand is still allowed but you're on your own.
* Add a way to combine the stdin and stdout into a single stream, as many times you just need the complete subprocess output.

Making the polling work on Windows is, as always, a bit challenging as `select()` doesn't work for pipes and neither do many other parts of the async IO API. A quick search shows that pretty much everyone is "emulating" select by polling (`PeekNamedPipe`) and sleeping... Not the most elegant solution but it seems to work.

**Comment by prime31** [comment]
I just ran into the deadlock on macos with a process that dumps a lot of data to stderr.

**Comment by LemonBoy** [comment]
Does #6750 work for you?

**Comment by andrewrk** [comment]
I did a partial merge of @LemonBoy's #6750 which solves the problem for the `exec` use case. The partial merge solves the issue for POSIX but not Windows. I'm closing the PR because it's an old draft, but this comment is a reminder that there is working(?) tested(?) code for Windows sitting around that could be utilized to solve this use case.

**Comment by marler8997** [comment]
I ran into this today.  I put together this program to reproduce on Windows:

```zig
const std = @import("std");

pub fn main() !void {
    const args = (try std.process.argsAlloc(std.heap.page_allocator))[1..];
    if (args.len == 0) {
        const self_exe_path = try std.fs.selfExePathAlloc(std.heap.page_allocator);
        std.log.info("executing child process '{s}'", .{self_exe_path});
        const result = try std.ChildProcess.exec(.{
            .allocator = std.heap.page_allocator,
            .argv = &[_][]const u8 { self_exe_path, "child"},
        });
        std.log.info("child process done", .{});
    } else {
        _ = try std.ChildProcess.exec(.{
            .allocator = std.heap.page_allocator,
            .argv = &[_][]const u8 { "bad_exe_does_not_exist"},
        });
    }
}
```

**Comment by LemonBoy** [comment]
Try applying the windows code path contained in #6750.

**Comment by marler8997** [comment]
@LemonBoy, tested and it fixes the issue :)

**Comment by andrewrk** [comment]
Fixed by @marler8997 and the fix landed in 0.8.1.
